### PR TITLE
Remove `has_peptide_quantifications`, deprecate associated classes and slots, and implement migrator

### LINKED ISSUE #2231: Re-instate full MetaProteomics records in validation
We should revert the changes implemented here: https://github.com/microbiomedata/berkeley-schema-fy24/pull/270 now that the MetaProteomics have been fixed (close this issue: https://github.com/microbiomedata/issues/issues/897).

### LINKED ISSUE #2250: Remove `ProteinQuantification` and `PeptideQuantification` classes and associated slots and enums.
According to [ADR #0017](https://github.com/microbiomedata/issues/pull/920), we've decided to remove the protein results data from mongo DB.

We will need to deprecate the `ProteinQuantification` and `PeptideQuantification` classes as well as their associated slots and enums and implement a migrator to update the data.

### PR DESCRIPTION
**Pull request by kheal**
Closes #2250 

This PR will

1) Remove the `has_peptide_quantifications` slot from the `MetaproteomicsAnalysis` class.
2) Implement a migrator to remove all `has_peptide_quantifications` data from `MetaproteomicsAnalysis` records.
3) Deprecate the class `ProteinQuantification` (range of the `has_peptide_quantifications`, no other application in the schema).
4) Deprecate the class `PeptideQuantification` (no connection in the schema's `Database` class).
5) Deprecate the following slots : `all_proteins`, `best_protein`, `min_q_value`, `peptide_sequence`, `peptide_spectral_count`, `peptide_sum_masic_abundance`, `peptide_sequence_count`, `protein_spectral_count`, `protein_sum_masic_abundance`.  The only application of these slots are on the now-deprecated `ProteinQuantification` and `PeptideQuantification` classes.
6) Update the makefile to remove a removal of `has_peptide_quantifications` data for the `make  local/mongo_as_unvalidated_nmdc_database.yaml` (Closes #2231, reverts https://github.com/microbiomedata/berkeley-schema-fy24/pull/270)

### CODE REVIEW

**Comment by kheal** [comment]
I've tested the migrator (in its combined form `migrator_from_11_0_3_to_11_1_0` ) and it passed validation with no errors.

File: nmdc_schema/migrators/partials/migrator_from_11_0_3_to_11_1_0/migrator_from_11_0_3_to_11_1_0_part_2.py
Comment (eecavanna) [inline nmdc_schema/migrators/partials/migrator_from_11_0_3_to_11_1_0/migrator_from_11_0_3_to_11_1_0_part_2.py:1]: Migrator looks great — thanks!

**Review by eecavanna — APPROVED** [review]
I reviewed the two files that involve migrators. Both look good to me and I'm comfortable with those files being merged in as is.

I didn't review the schema or Makefile changes, and don't plan to.

**Comment by eecavanna** [comment]
Hi @turbomam, the deadline for getting this into the schema is tomorrow, according to the [timeline of events](https://github.com/microbiomedata/issues/issues/914) for the upcoming release (scheduled for next Tuesday).

Will you take a look at this PR today; in particular, the files outside of the `migrators/` directory?

File: project.Makefile
Comment (eecavanna) [inline project.Makefile:1]: The changes in this file make sense to me given that the `has_peptide_quantifications` slot is being removed in this PR.

File: nmdc_schema/migrators/partials/migrator_from_11_0_3_to_11_1_0/migrator_from_11_0_3_to_11_1_0_part_3.py
Comment (eecavanna) [inline nmdc_schema/migrators/partials/migrator_from_11_0_3_to_11_1_0/migrator_from_11_0_3_to_11_1_0_part_3.py:10]: Since this is not the first partial migrator in the sequence, I suggest updating this to identify the _preceding_ partial migrator.

```suggestion
    _from_version = "11.1.0.part_2"
```

Comment (eecavanna) [reply nmdc_schema/migrators/partials/migrator_from_11_0_3_to_11_1_0/migrator_from_11_0_3_to_11_1_0_part_3.py:10]: I will go ahead and commit this change since it's after hours and the lead repo maintainer is in a timezone that is a few hours ahead of the other contributors (and so may start processing this PR before the other contributors have begun work for the day).

Y'all can undo this change if you want to, after that.

**Comment by turbomam** [comment]
> @kheal and @eecavanna is `PeptideQuantification` actually deprecated in this PR? I do see that `ProteinQuantification` is deprecated.

I found it!

**Comment by turbomam** [comment]
You can double check with @mslarae13 and or @sierra-moxon, but I think we remove the assignments of deprecated slots in classes. Like I would expect something like this:

```yaml
  PeptideQuantification:
    class_uri: nmdc:PeptideQuantification
    description: This is used to link a metaproteomics analysis workflow to a specific peptide sequence and related information
    slots:
      - type
    deprecated: "not used. 2024-11 https://github.com/microbiomedata/nmdc-schema/issues/2250"
```

**Comment by eecavanna** [comment]
Took me a while to find, too, given the way GitHub's diff viewer happened to collapse things. Here's a screnshot:

image

**Comment by turbomam** [comment]
Do we know what happens if we include an instantiation of a deprecated class in the invalid examples file directory?

**Review by turbomam — COMMENTED** [review]
See comments about removing deprecated slots from classes and including invalid examples that instantiates the deprecated classes, as long as that doesn't cause an error

**Comment by eecavanna** [comment]
Here are the [deprecation guidelines](https://github.com/microbiomedata/nmdc-schema/blob/a5c3f4e7fe221b01d24151e536bfe00cbeb3ac19/src/docs/schema_element_deprecation_guide.md#L4) I see. Deprecation happens in two stages (i.e. across two consecutive releases). In the first release (what we're working on now), the guidance is to "add deprecation-related annotations" to the element's definition. In the second release (that'll be next month), one of the multiple pieces of guidance is to remove references to the deprecated elements.

**Comment by kheal** [comment]
I've added an invalid example for the `ProteinQuantification` class.  They other deprecated class (`PeptideQuantification`) is in-line only and its associated slot has been removed.  Therefore any invalid examples with this class would be invalid for more than one reason.

I see value in keeping the deprecated slots on a deprecated class. Its allows for transparency for why the slots ever existed.  By removing those slots from the class, we remove that transparency.

I will add (and link) an issue to move the deprecated aspects from this release to the deprecated.yaml file for the next release.

**Comment by turbomam** [comment]
Thanks for sharing your reasoning for leaving the associated classes, @kheal 

The [deprecation guidelines](https://github.com/microbiomedata/nmdc-schema/blob/a5c3f4e7fe221b01d24151e536bfe00cbeb3ac19/src/docs/schema_element_deprecation_guide.md#L4) shared by @eecavanna do say to remove them.

> **Remove any references to the deprecated element from the original schema YAML file. **
> 
> ** note: For a deprecated class, references could be in the is_a or mixin elements of another class, or in the domain or range elements of an existing slot. For deprecated slots, check all classes that reference that slot, etc.

**Comment by eecavanna** [comment]
@turbomam, that guidance is in the "Second release cycle" section.

image

We are in the first release cycle here.

**Comment by turbomam** [comment]
Thanks guys, I got my phases wrong.

I have to think about the invalid examples some more. @kheal does have a good point about sticking to the practice of including a single invalid feature.

**Comment by eecavanna** [comment]
Sounds good. Is that thinking something you will be doing before merging this in, or after? In other words, is there anything more you expect anyone to do in this PR before it gets merged in?

**Comment by turbomam** [comment]
I don't think we have thought through the implications of the deprecation process on the invalid examples as well as we have through the rest of the deprecation process.

`src/data/invalid/PeptideQuantification_deprecated.yaml`:

```yaml
# This example is invalid because the ProteinQuantification type is deprecated.
type: nmdc:PeptideQuantification
```

that doesn't fail the validation framework, ie it doesn't fail to fail to validate there, but it does fail to fail to fail when directly using `linkml-validate`

```shell
poetry run linkml-validate \
    --schema nmdc_schema/nmdc_materialized_patterns.yaml \
    --target-class ProteinQuantification src/data/invalid/ProteinQuantification_deprecated.yaml 
```

> No issues found

So I am going to walk back my request for invalid example files. Sorry bout that, @kheal . Do you want to delete it or should I?

**Comment by eecavanna** [comment]
As I said on Slack just now, I'll delete the deprecation-specific example file from this branch. I'll do it now.

**Review by turbomam — APPROVED** [review]

**Pull request merged**
